src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,333 @@
+            : this()
+        {
+            if (functionSets == null)
+            {
+                throw new ArgumentNullException($"{nameof(functionSets)} cannot be null", nameof(functionSets));
+            }
+
+            foreach (var functionSet in functionSets)
+            {
+                Add(functionSet);
+            }
+        }
+
+        // Slow API, only use for backward compatibility
+        [Obsolete("Slow API. Prefer using With* functions for identifying functions you need.")]
+        internal IEnumerable<TexlFunction> Functions
+        {
+            get
+            {
+                foreach (var kvp in _functions)
+                {
+                    foreach (var func in kvp.Value)
+                    {
+                        yield return func;
+                    }
+                }
+            }
+        }
+
+        internal TexlFunction Append(TexlFunction function) => Add(function);
+
+        internal TexlFunction Add(TexlFunction function)
+        {
+            if (function == null)
+            {
+                throw new ArgumentNullException($"{nameof(function)} cannot be null", nameof(function));
+            }
+
+            var fList = WithNameInternal(function.Name);
+
+            if (fList.Any())
+            {
+                if (fList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions");
+                }
+
+                fList.Add(function);
+            }
+            else
+            {
+                _functions.Add(function.Name, new List<TexlFunction>() { function });
+            }
+
+            var fInvariantList = WithInvariantNameInternal(function.LocaleInvariantName);
+
+            if (fInvariantList.Any())
+            {
+                if (fInvariantList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions (invariant)");
+                }
+
+                fInvariantList.Add(function);
+            }
+            else
+            {
+                _functionsInvariant.Add(function.LocaleInvariantName, new List<TexlFunction>() { function });
+            }
+
+            var fnsList = WithNamespaceInternal(function.Namespace);
+
+            if (fnsList.Any())
+            {
+                if (fnsList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions (namespace)");
+                }
+
+                fnsList.Add(function);
+            }
+            else
+            {
+                _namespaces.Add(function.Namespace, new List<TexlFunction>() { function });
+            }
+
+            _enums.AddRange(function.GetRequiredEnumNames());
+
+            _count++;
+
+            return function;
```

Comment:
>return [](http://example.com/codeflow?start=12&length=6)

could we add a Contract.Assert that count == _enums.Count == _namespaces.Count , etc

Reply:
Not possible due to overloads or it's going to be an expensive check